### ISSUE TITLE
EPIC: get TrOCR training actually running — the backend exists, no model has ever been trained with it

### ISSUE THREAD

**Issue by thodel** [thodel/serving-atr-inference#99]
**The backend already exists.** `engines/trocr_train_svc/` has `runner.py` (218 lines),
`train_trocr.py`, `evaluate_trocr.py` and `requirements.txt`; `trocr` is registered in
`BACKENDS`; `tests/test_trocr_train_pipeline.py` and `tests/test_training_trocr_cmd.py`
cover it; and `.venvs/trocr-train` is built on the box. A `POST /train/jobs` with
`{"engine": "trocr", …}` is accepted today.

**What is missing is that it has never run.** `trained_root` holds kraken and vllm
models only — no TrOCR model has ever been produced, so nothing in that path has met a
GPU, a real dataset or a real tokenizer. This epic is about closing that gap, not
about building the backend.

Why it matters now: we are assembling a **multilingual** corpus (FoNDUE-GD, 26
reachable sources, fr/de/es/la/nl, ~20,700 pages / 660,000 lines). TrOCR is an
encoder-decoder with a *language model* in the decoder, which is exactly where a CTC
model has nothing — and equally exactly where a monolingual decoder on multilingual
data goes wrong. Ströbel's own numbers have TrOCR at the top of the table (F1 0.990
against kraken+ 0.954), and our VLM run reached CER 0.2324 where kraken managed 0.3921
on the same material. Both say the same thing: the autoregressive family deserves a
real attempt here.

## T1 — Make the venv real
- [ ] Build/refresh `.venvs/trocr-train` from `requirements.txt` and record the resolved
      `transformers`/`torch` versions. kraken pins `datasets<4` and its own transformers
      range; TrOCR must not be allowed to drift into the same tree (`docs/VLM_TRAINING.md`
      §"Why one service and two venvs").
- [ ] `python -m trocr_train_svc.runner --help` in that venv — the smoke test that the
      module named in `BACKENDS` actually exists there (#44 was exactly this failure).

## T2 — Choose the base model, and say why
- [ ] `microsoft/trocr-base-handwritten` is English-only in both tokenizer and decoder.
      For fr/de/es/la/nl that is the wrong starting point, and the decoder's LM prior
      will fight the data.
- [ ] Evaluate the alternatives on a few hundred lines before committing a run:
      `microsoft/trocr-large-handwritten`, a multilingual-decoder variant, or an
      encoder-only start with a fresh multilingual tokenizer.
- [ ] **Record tokenizer coverage against the corpus charset.** A decoder that cannot
      emit a character can never learn it — the same structural ceiling that leaves all
      three kraken models at 0/170 on combining marks. Measure it, do not assume it.

## T3 — First real run (smoke, then scale)
- [ ] ~2,000 lines, 1 epoch, on the material we already have compiled. Success is "a
      CER that is not 1.0", nothing more — the VLM smoke run (0.466 against a base of
      1.837) is the precedent for what this stage proves.
- [ ] Then a full run on the FoNDUE corpus once it is prepared.
- [ ] Convergence guard: `FLOOR_BY_ENGINE` in `convergence.py` has an entry for `vllm`
      (25 steps, measured) and none for `trocr`. Fine-tuning a pretrained
      encoder-decoder is closer to the VLM case than to from-scratch CTC; set the floor
      from the smoke run rather than by analogy.

## T4 — Comparable evaluation
- [ ] Score against the **same** held-out set as kraken, through `textmetrics.score_pairs`,
      so a TrOCR CER and a kraken CER remain the same kind of number (that is already
      the contract — `docs/VLM_TRAINING.md`).
- [ ] Report the error *shape*, not just the CER. TrOCR fails by not stopping at the
      line (the VLM run's signature: deletions dominant under this project's inverted
      convention), CTC by blank collapse. A single number cannot distinguish them.

## T5 — Serving
- [ ] `trocr_svc` blocks its event loop (#95), so a trained TrOCR model cannot be served
      concurrently even once it exists. Fix before promoting anything, or the promotion
      gate (#36) will pass a model that then serves badly under load.

## Not in scope
Multi-GPU, distillation, and the decoder-LM questions that only become interesting once
a first honest CER exists.

**Comment by thodel** [comment]
## T1 erledigt (`0b2e5e4`)

`python -m trocr_train_svc.runner --help` läuft jetzt im eigenen venv. Zwei echte
Defekte auf dem Weg, beide erst **in der train-Stufe** sichtbar — nach prepare und
compile:

* **`accelerate` fehlte in `requirements.txt`.** `from transformers import
  Seq2SeqTrainer` gelingt ohne; der `ImportError` fällt erst, wenn
  `Seq2SeqTrainingArguments` seine Devices auflöst.
* **`pydantic`, `pydantic-settings`, `pyyaml`, `Pillow` fehlten.** Der Runner führt die
  geteilte `BasePipeline` aus und braucht denselben Kern wie jedes andere Backend, nicht
  nur den Modell-Stack. kraken-train und vlm-train listen sie; trocr-train nicht.

`scripts/check_requirements.py` (#53) konnte das nicht fangen — es prüft installiert
gegen Requirements-Datei, und die Datei war die Lücke. Jetzt exit 0 und durchsetzbar.

## T2: die Tokenizer-These war falsch

Gemessen statt vermutet — Round-Trip über 400 echte Zeilen und das ganze Korpusalphabet:

| | trocr-base-handwritten | trocr-large-handwritten |
|---|---|---|
| Typ | RobertaTokenizerFast, 50.265 | identisch |
| Zeichen ohne Round-Trip | **0 / 77** | **0 / 77** |
| Zeilen exakt zurück | **400 / 400** | 400 / 400 |
| Tokens je Zeichen | 0.383 | 0.383 |

**Der Grund ist byte-level BPE**: der Tokenizer fällt auf Bytes zurück und kann damit
jedes Unicode-Zeichen darstellen. Die strukturelle Decke, die ich hier vermutet hatte —
und die kraken bei den 170 kombinierenden Zeichen tatsächlich hat — existiert für TrOCR
nicht. Der Punkt in T2 oben, "a decoder that cannot emit a character can never learn
it", trifft auf diese Modellfamilie nicht zu.

Was bleibt, ist ein **Prior**, kein Repräsentationsproblem: ein auf Englisch
vortrainierter Sprachmodell-Decoder bevorzugt englische Fortsetzungen. Das ist durch
Fine-Tuning behebbar, anders als eine fehlende Zeichenklasse — und es senkt das Risiko
von `trocr-base-handwritten` als Startpunkt erheblich. 0.383 Tokens je Zeichen (~2,6
Zeichen pro Token) auf flämischem Text spricht zusätzlich dagegen, dass die
Segmentierung entgleist.

**Einschränkung**: gemessen auf `val_clean` (flämisch, 77 Zeichen, **kein einziges**
kombinierendes Zeichen — die 170 `Inherited` stecken im Test-, nicht im
Trainingsalphabet). Für das FoNDUE-Korpus (lat/de mit Kürzungszeichen, fr/es/nl) ist
der Test zu wiederholen, sobald der Download durch ist. Erst dann ist die
Basismodellwahl wirklich entschieden.

### Angepasste T2-Aufgaben

- [x] Tokenizer-Abdeckung gegen den Korpus-Zeichensatz messen — für flämisch erledigt,
      Ergebnis: vollständig
- [ ] Wiederholen auf dem FoNDUE-Alphabet, sobald vorbereitet
- [ ] Basismodell: `trocr-base-handwritten` als Default vorsehen, `large` als Kandidat
      für rung 1. Der frühere Einwand "englischer Tokenizer" ist erledigt; offen bleibt
      nur der Decoder-Prior, und den entscheidet ein Lauf, keine Analyse.

**Comment by thodel** [comment]
## T3 angefangen, dabei sieben Fehler gefunden — der Trainer war nie gelaufen

`4d78c34` auf `main`. Der Ablauf, weil die Reihenfolge der Fehler selbst der Befund ist.

Der erste echte Aufruf von `trocr_train_svc` (Smoke-Job
`20260902T071101Z-trocr-thun-smoke-v1`, Thun-Satz, lokaler Job-Store weil der Share
damals offline war) brach ab. Nach dem Fix brach er an der nächsten Stelle ab. Und so
weiter — **sieben Mal**, jeder Fehler erst sichtbar, nachdem der vorige behoben war:

| # | Fehler | Wirkung |
|---|---|---|
| 1 | `length_penalty` ist in transformers 4.57 kein Feld von `Seq2SeqTrainingArguments` | TypeError beim Bau des Trainers |
| 2 | `tokenizer=` wurde aus `Trainer.__init__` entfernt | TypeError |
| 3 | Trainer mit `model=None` gebaut, Modell danach geladen und **nie zugewiesen** | hätte mit None trainiert |
| 4 | `resize_token_embeddings` wirft auf `VisionEncoderDecoderModel` | NotImplementedError |
| 5 | `length_penalty` bei `num_beams=1` wird als ungültig abgelehnt | ValueError |
| 6 | Trainer entfernt Dataset-Spalten, die `forward` nicht kennt — also `image_path` und `text` | KeyError im ersten Batch |
| 7 | `TrOCRProcessor` liefert `pixel_values` + `labels`, kein `input_ids` | KeyError im Collator |

Die Fehler 1, 2, 4, 5 hängen an transformers 4.57; 3, 6, 7 wären in jeder Version
aufgetreten. Der Code hatte **35 grüne Tests** — keiner davon fasst die echte
transformers-API an, deshalb hat keiner davon irgendetwas gefunden.

### Verifiziert, nicht nur repariert

Eine Epoche über vier synthetische Zeilenbilder auf der CPU, inklusive Evaluation mit
`predict_with_generate` und geschriebenem Checkpoint:

```
{'eval_loss': 4.17, 'epoch': 1.0}
{'train_runtime': 7.31, 'train_loss': 7.88, 'epoch': 1.0}
checkpoint saved to .../out
```

Die Zahlen sind bedeutungslos (vier Bilder), der Durchlauf ist der Punkt.

**`scripts/smoke_trocr.py`** macht genau das wiederholbar — ~30 Sekunden, ohne GPU, ohne
Datensatz, ohne Vorbereitung. Bewusst ein Skript und kein weiterer Test mit Attrappen:
Attrappen haben hier siebenmal nichts gemerkt.

Zwei Runden gingen dabei für Fehler in meinem eigenen Prüf-Harness drauf, nicht im Code:
`microsoft/trocr-small-handwritten` liefert über `AutoProcessor` einen nackten Tokenizer
statt eines `TrOCRProcessor` (das konfigurierte `trocr-base-handwritten` nicht), und mein
JSONL benutzte `image_path` statt `image`. Erwähnt, weil beides beim nächsten Mal wieder
Zeit kosten würde.

## Stand von T3: **nicht erledigt**

Es ist **kein TrOCR-Modell trainiert worden**. Was fehlt:

- [ ] Der Smoke-Job muss neu laufen. Der alte Job-Record steht auf `failed`.
- [ ] **Die Box hat den Fix nicht.** `~/Repo/serving-atr-inference` steht auf `87d1811`,
      zehn Commits hinter `main`, und `git pull` bricht an einer unversionierten Änderung
      in `engines/trocr_svc/app.py` ab, die nicht von mir stammt. Solange die nicht
      committet oder gestasht ist, läuft dort weiterhin die kaputte Fassung — ein neuer
      Lauf würde direkt in Fehler 1 laufen.
- [ ] Danach: Smoke gegen den Thun-Satz, Erfolgskriterium bleibt „ein CER, der nicht 1.0
      ist", und der Vergleich gegen kraken (0.9838 from scratch, 0.3921 fine-tune) und
      VLM (0.466) auf demselben Material.

## Nachtrag zum Eval-Material

Seit der Epic-Formulierung gibt es ein **deutsches, dokumentgetrenntes Held-out-Set**
(#98): `german_test.arrow`, 695 Seiten aus 200 Dokumenten, 882.255 Zeichen. Das ist ab
jetzt die richtige Messlatte für ein TrOCR-Modell auf deutschem Material — nicht der
Thun-Satz, der nur für den Vergleich mit den drei alten Läufen taugt.

**Comment by thodel** [comment]
The training part of this epic continues in https://github.com/thodel/training-atr-models/issues/37 (the code lives in training-atr-models since the split). T5 (serving side: #95, and trocr_svc loading a registered local_path) stays here; the training work T2–T4 continues in https://github.com/thodel/training-atr-models/issues/37.
